Match ergonomics 2024: let `&` patterns eat `&mut`

Jules-Bertholet opened the pull request:
r? @Nadrieril

cc https://github.com/rust-lang/rust/issues/123076

@rustbot label A-edition-2024 A-patterns

Jules-Bertholet committed:
Match ergonomics 2024: let `&` patterns eat `&mut`

Nadrieril commented on compiler/rustc_hir_typeck/src/pat.rs:
```diff
@@ -2125,45 +2185,65 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                                PatInfo {
+                                    max_ref_mutbl: pat_info
+                                        .max_ref_mutbl
+                                        .cap_mutbl_to_not(Some(pat.span.until(inner.span))),
```
Unrelatedly, I think you could cap `pat_info.max_ref_mutbl` regardless of `expected`, which would be a bit easier to follow. More generally I'm surprised that you do this capping dance in two separate places; do you think there's a way to do this only once?

Jules-Bertholet replied:
> I am worried that this `until` would break down in macros.

It does. What should I be doing instead?

Nadrieril replied:
I looked through the `Span` docs, my best guess is `pat.span.until(inner.span.find_ancestor_inside_same_ctxt(pat.span))` (maybe you have to swap arguments? I'm not sure, doc is here: [find_ancestor_inside_same_ctxt](https://doc.rust-lang.org/nightly/nightly-rustc/rustc_span/span_encoding/struct.Span.html#method.find_ancestor_inside_same_ctxt))

Nadrieril commented on compiler/rustc_hir_typeck/src/pat.rs:
```diff
@@ -692,8 +736,22 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                E0596,
+                "cannot bind with `ref mut` behind an `&` pattern"
+            );
+            err.span_help(and_pat_span, "change this `&` pattern to an `&mut`");
```
How come this doesn't trigger on `let &ref mut x = &0`?

Jules-Bertholet replied:
We only want to trigger the error in cases where the suggestion of replacing an `&` pattern with `&mut` would actually work. If the type being matched against contains a shared reference, no modifications to the pattern will allow you to get `&mut`, so we don't show the suggestion and instead let the move-checker find and report the error.

Nadrieril replied:
I see. The other change I just suggested to `MutblCap` will clarify this

Jules-Bertholet committed:
Various fixes:

- Only show error when move-check would not be triggered
- Add structured suggestion

Nadrieril commented on compiler/rustc_hir_typeck/src/pat.rs:
```diff
@@ -2185,7 +2196,21 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
+                                pat_info
+                            };
+                            (expected, r_ty, pat_info)
+                        }
```
I am confused by this, does it do what you want? I meant to suggest to move the whole `let pat_info = if r_mutbl == Mutability::Not ...` before the match.

Nadrieril replied:
Oh, I think I see what you're doing. Is this only to disable diagnostics? Please add a comment explaining that the mutability is already capped but you disable diagnostics to let the borrow-checker catch this error. Or why not catch the error yourself? You're not doing a worse job than the borrow-checker here, and you have more information for suggestions.

Also, I'm pretty sure you can do this in `calc_default_binding_mode` instead, which I would prefer.

Nadrieril commented on compiler/rustc_hir_typeck/src/pat.rs:
```diff
@@ -165,7 +165,13 @@ enum MutblCap {
 impl MutblCap {
     fn cap_mutbl_to_not(self, span: Option<Span>) -> Self {
-        if self == MutblCap::Mut { MutblCap::Not(span) } else { self }
+        if let Some(s) = span
+            && self != MutblCap::Not(None)
+        {
+            MutblCap::Not(Some(s))
+        } else {
+            MutblCap::Not(None)
+        }
```
It seems like whether or not the `Span` is present is important in some way. Can you document exactly what this means?

Nadrieril commented on compiler/rustc_hir_typeck/src/pat.rs:
```diff
@@ -746,9 +752,14 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
-                "cannot bind with `ref mut` behind an `&` pattern"
+                "cannot borrow as mutable inside an `&` pattern"
+            );
+            err.span_suggestion(
+                and_pat_span,
+                "replace this `&` with `&mut`",
+                "&mut ",
+                Applicability::MachineApplicable,
```
Could you add a comment explaining why this can only happen when the type of the pattern is `&mut T`? At first glance I'd expect it to trigger also in the `let &mut ref x = &0;` case.

Jules-Bertholet committed:
Fix spans when macros are involved

rust-log-analyzer commented:
The job **`x86_64-gnu-llvm-17`** failed! Check out the build log: [(web)](https://github.com/rust-lang/rust/actions/runs/8952736249/job/24590449932) [(plain)](https://github.com/rust-lang/rust/commit/9605d90b7b5f2ef1ac15788d156907a80589bdbd/checks/24590449932/logs)

<details><summary><i>Click to see the possible cause of the failure (guessed by this bot)</i></summary>

```plain
#16 exporting to docker image format
#16 sending tarball 29.8s done
#16 DONE 35.1s
##[endgroup]
Setting extra environment values for docker:  --env ENABLE_GCC_CODEGEN=1 --env GCC_EXEC_PREFIX=/usr/lib/gcc/
[CI_JOB_NAME=x86_64-gnu-llvm-17]
---
sccache: Starting the server...
##[group]Configure the build
configure: processing command line
configure: 
configure: build.configure-args := ['--build=x86_64-unknown-linux-gnu', '--llvm-root=/usr/lib/llvm-17', '--enable-llvm-link-shared', '--set', 'rust.thin-lto-import-instr-limit=10', '--set', 'change-id=99999999', '--enable-verbose-configure', '--enable-sccache', '--disable-manage-submodules', '--enable-locked-deps', '--enable-cargo-native-static', '--set', 'rust.codegen-units-std=1', '--set', 'dist.compression-profile=balanced', '--dist-compression-formats=xz', '--disable-dist-src', '--release-channel=nightly', '--enable-debug-assertions', '--enable-overflow-checks', '--enable-llvm-assertions', '--set', 'rust.verify-llvm-ir', '--set', 'rust.codegen-backends=llvm,cranelift,gcc', '--set', 'llvm.static-libstdcpp', '--enable-new-symbol-mangling']
configure: target.x86_64-unknown-linux-gnu.llvm-config := /usr/lib/llvm-17/bin/llvm-config
configure: llvm.link-shared     := True
configure: rust.thin-lto-import-instr-limit := 10
configure: change-id            := 99999999
---
---- [ui] tests/ui/let-else/let-else-drop-order.rs stdout ----

error: test compilation failed although it shouldn't!
status: exit status: 101
command: env -u RUSTC_LOG_COLOR RUSTC_ICE="0" RUST_BACKTRACE="short" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2/bin/rustc" "/checkout/tests/ui/let-else/let-else-drop-order.rs" "-Zthreads=1" "-Zsimulate-remapped-rust-src-base=/rustc/FAKE_PREFIX" "-Ztranslate-remapped-path-to-local-path=no" "-Z" "ignore-directory-in-diagnostics-source-blocks=/cargo" "-Z" "ignore-directory-in-diagnostics-source-blocks=/checkout/vendor" "--sysroot" "/checkout/obj/build/x86_64-unknown-linux-gnu/stage2" "--target=x86_64-unknown-linux-gnu" "--check-cfg" "cfg(FALSE)" "-O" "--error-format" "json" "--json" "future-incompat" "-Ccodegen-units=1" "-Zui-testing" "-Zdeduplicate-diagnostics=no" "-Zwrite-long-types-to-disk=no" "-Cstrip=debuginfo" "-C" "prefer-dynamic" "-o" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/let-else/let-else-drop-order/a" "-A" "internal_features" "-Crpath" "-Cdebuginfo=0" "-Lnative=/checkout/obj/build/x86_64-unknown-linux-gnu/native/rust-test-helpers" "-L" "/checkout/obj/build/x86_64-unknown-linux-gnu/test/ui/let-else/let-else-drop-order/auxiliary" "--edition=2021"
--- stderr -------------------------------
thread 'rustc' panicked at compiler/rustc_hir_typeck/src/pat.rs:438:139:
called `Option::unwrap()` on a `None` value
stack backtrace:
stack backtrace:
   0: rust_begin_unwind
   1: core::panicking::panic_fmt
   2: core::panicking::panic
   3: core::option::unwrap_failed
   4: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_pat
   5: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_decl
   6: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_block_with_expected
   7: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_kind
   8: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
   9: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation
  10: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_block_with_expected
  11: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_kind
  12: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  13: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation
  14: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_block_with_expected
  15: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_kind
  16: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation_and_args
  17: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_expr_with_expectation
  18: <rustc_hir_typeck::fn_ctxt::FnCtxt>::check_return_expr
  19: rustc_hir_typeck::check::check_fn
  20: rustc_hir_typeck::typeck
      [... omitted 2 frames ...]
  21: <rustc_middle::hir::map::Map>::par_body_owners::<rustc_hir_analysis::check_crate::{closure#4}>::{closure#0}
  22: std::panicking::try::<(), core::panic::unwind_safe::AssertUnwindSafe<rustc_data_structures::sync::parallel::enabled::par_for_each_in<&rustc_span::def_id::LocalDefId, &[rustc_span::def_id::LocalDefId], <rustc_middle::hir::map::Map>::par_body_owners<rustc_hir_analysis::check_crate::{closure#4}>::{closure#0}>::{closure#0}::{closure#1}::{closure#0}>>
  23: <rustc_data_structures::sync::parallel::ParallelGuard>::run::<(), rustc_data_structures::sync::parallel::enabled::par_for_each_in<&rustc_span::def_id::LocalDefId, &[rustc_span::def_id::LocalDefId], <rustc_middle::hir::map::Map>::par_body_owners<rustc_hir_analysis::check_crate::{closure#4}>::{closure#0}>::{closure#0}::{closure#1}::{closure#0}>
  25: rustc_interface::passes::analysis
  25: rustc_interface::passes::analysis
      [... omitted 2 frames ...]
  27: <rustc_interface::interface::Compiler>::enter::<rustc_driver_impl::run_compiler::{closure#0}::{closure#1}, core::result::Result<core::option::Option<rustc_interface::queries::Linker>, rustc_span::ErrorGuaranteed>>
  28: <scoped_tls::ScopedKey<rustc_span::SessionGlobals>>::set::<rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#0}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#0}::{closure#0}::{closure#0}, core::result::Result<(), rustc_span::ErrorGuaranteed>>
  29: rustc_span::create_session_globals_then::<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_interface::util::run_in_thread_with_globals<rustc_interface::util::run_in_thread_pool_with_globals<rustc_interface::interface::run_compiler<core::result::Result<(), rustc_span::ErrorGuaranteed>, rustc_driver_impl::run_compiler::{closure#0}>::{closure#1}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#0}, core::result::Result<(), rustc_span::ErrorGuaranteed>>::{closure#0}::{closure#0}::{closure#0}>
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
---
note: please make sure that you have updated to the latest nightly

note: rustc 1.80.0-nightly (3642d479d 2024-05-04) running on x86_64-unknown-linux-gnu

note: compiler flags: -Z threads=1 -Z simulate-remapped-rust-src-base=/rustc/FAKE_PREFIX -Z translate-remapped-path-to-local-path=no -Z ignore-directory-in-diagnostics-source-blocks=/cargo -Z ignore-directory-in-diagnostics-source-blocks=/checkout/vendor -C codegen-units=1 -Z ui-testing -Z deduplicate-diagnostics=no -Z write-long-types-to-disk=no -C strip=debuginfo -C prefer-dynamic -C rpath -C debuginfo=0
query stack during panic:
query stack during panic:
#0 [typeck] type-checking `main`
end of query stack
------------------------------------------


```

</details>

Jules-Bertholet committed:
Comments and fixes

Jules-Bertholet committed:
Rename `explicit_ba`

Nadrieril commented on compiler/rustc_hir_typeck/src/pat.rs:
```diff
@@ -436,7 +436,7 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
             //
             // See issue #46688.
-            PatKind::Ref(inner, mutbl) => AdjustMode::ResetAndConsumeRef(*mutbl, inner.span),
+            PatKind::Ref(inner, mutbl) => AdjustMode::ResetAndConsumeRef(*mutbl, pat.span.until(inner.span.find_ancestor_inside(pat.span).unwrap())),
```
The documentation for `find_ancestor_inside` explicitly says to not use it to expand the span, which I understood would probably apply to `until` too. Did `find_ancestor_inside_same_ctxt` not do what you want?

Jules-Bertholet replied:
I am shrinking the span, not expanding it?

Nadrieril replied:
`until` doesn't know that though. I don't really understand what `SyntaxContext`s are but it sounded good to use the same one?

Jules-Bertholet replied:
Looking at the `rustc_span` source code, I think it's fine, because `until` will use the `SyntaxContext` of the first (outer) span anyway—so anything weird about the `SyntaxContext` of the span passed as the last argument to `until` won't influence the result. (IOW, we are not "extending the *returned* span", we are shrinking a different span that is not the returned span.)

Nadrieril replied:
Ok, that's fine then

Nadrieril commented on compiler/rustc_hir_typeck/src/pat.rs:
```diff
@@ -156,15 +156,23 @@ enum AdjustMode {
-    /// Mutability restricted to immutable;
-    /// contained span, if present, should be shown in diagnostics as the reason.
-    Not(Option<Span>),
+    /// Mutability restricted to immutable.
+    ///
+    /// The contained span, if present, points to an `&` pattern
+    /// that is the reason for the restriction,
+    /// and which will be reported in a diagnostic.
+    /// (Said diagnostic is shown only if
+    /// replacing the `&` pattern with `&mut` would allow the code to compile.)
+    ///
+    /// (Outer [`Option`] is for whether to show the diagnostic,
+    /// inner [`Option`] is for whether we have a span we can report)
+    Not(Option<Option<Span>>),
```
I am really not a fan of `Option<Option<`. Could you make the outer `Option` into a different variant of `MutblCap`? And split `cap_mutbl_to_not` into two functions to match

Jules-Bertholet committed:
No more `Option<Option<>>`

Nadrieril commented on compiler/rustc_hir_typeck/src/pat.rs:
```diff
@@ -157,34 +157,33 @@ enum AdjustMode {
-    /// (Outer [`Option`] is for whether to show the diagnostic,
-    /// inner [`Option`] is for whether we have a span we can report)
-    Not(Option<Option<Span>>),
+    WeaklyNot(Option<Span>),
```
Nice, good name :+1:

Nadrieril commented on compiler/rustc_hir_typeck/src/pat.rs:
```diff
@@ -157,34 +157,33 @@ enum AdjustMode {
     /// that is the reason for the restriction,
     /// and which will be reported in a diagnostic.
     /// (Said diagnostic is shown only if
     /// replacing the `&` pattern with `&mut` would allow the code to compile.)
```
That always applies now, doesn't it?

Jules-Bertholet committed:
Remove redundant comment

Nadrieril commented on compiler/rustc_hir_typeck/src/pat.rs:
```diff
@@ -2125,45 +2204,67 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
-                                    .insert(pat.hir_id);
-                                (expected, expected)
+                        ty::Ref(_, r_ty, r_mutbl) if r_mutbl == mutbl => {
+                            let pat_info = if r_mutbl == Mutability::Not
+                                && ((pat.span.at_least_rust_2024()
+                                    && self.tcx.features().ref_pat_eat_one_layer_2024)
+                                    || self.tcx.features().ref_pat_everywhere)
+                            {
+                                PatInfo { max_ref_mutbl: MutblCap::Not, ..pat_info }
```
Could this be done in `calc_default_binding_mode`? I think you have all the info required there

Jules-Bertholet replied:
Hmm… `check_pat_ref` uses `shallow_resolve` to get the reference type, but other parts of the file use `try_structurally_resolve_type` instead. Do you know what the difference is? Does it matter which one I use?

(Also, I'm not sure moving to `calc_default_binding_mode` would be cleaner. Right now, `calc_default_binding_mode` manages inherited references and `check_pat_ref` manages non-inherited references, which is a reasonably clean separation. Maybe we could do a wider refactor where `check_pat_ref` assumes responsibility for all `&`/`&mut` pattern logic, but that is best left to a future PR I think.)

Nadrieril replied:
I don't know the difference but I don't think it matters here. The reason I think it's better over on `calc_default_binding_mode` is that it's `max_ref_mutbl` capping happens there already so it would be easier to track in one place. In fact I think this one (line 2213) partially overrides the one on line 359

Jules-Bertholet replied:
>  a wider refactor where `check_pat_ref` assumes responsibility for all `&`/`&mut` pattern logic

I did this.

Nadrieril replied:
That's great, ty

Jules-Bertholet committed:
Move all ref pat logic into `check_pat_ref`

bors commented:
:umbrella: The latest upstream changes (presumably #124961) made this pull request unmergeable. Please [resolve the merge conflicts](https://rustc-dev-guide.rust-lang.org/git.html#rebasing-and-conflicts).

Jules-Bertholet committed:
Add comment on `cap_to_weakly_not`

Co-authored-by: Guillaume Boisseau <Nadrieril@users.noreply.github.com>

Nadrieril commented:
Ty!

@bors r+

bors commented:
:pushpin: Commit 6d5c6f541f5dc9f3142c84bbc5203e57c0a4a18c has been approved by `Nadrieril`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Nadrieril 6d5c6f541f5dc9f3142c84bbc5203e57c0a4a18c -->
<!-- homu: {"type":"Approved","sha":"6d5c6f541f5dc9f3142c84bbc5203e57c0a4a18c","approver":"Nadrieril","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

bors commented:
:hourglass: Testing commit 6d5c6f541f5dc9f3142c84bbc5203e57c0a4a18c with merge 35c5e67c69cbde49b47fe537e296803b6a25b456...
<!-- homu: {"type":"BuildStarted","head_sha":"6d5c6f541f5dc9f3142c84bbc5203e57c0a4a18c","merge_sha":"35c5e67c69cbde49b47fe537e296803b6a25b456"} -->

bors commented:
:sunny: Test successful - [checks-actions](https://github.com/rust-lang-ci/rust/actions/runs/9043089432/job/24851713364)
Approved by: Nadrieril
Pushing 35c5e67c69cbde49b47fe537e296803b6a25b456 to master...
<!-- homu: {"type":"BuildCompleted","approved_by":"Nadrieril","base_ref":"master","builders":{"checks-actions":"https://github.com/rust-lang-ci/rust/actions/runs/9043089432/job/24851713364"},"merge_sha":"35c5e67c69cbde49b47fe537e296803b6a25b456"} -->

rust-timer commented:
Finished benchmarking commit (35c5e67c69cbde49b47fe537e296803b6a25b456): [comparison URL](https://perf.rust-lang.org/compare.html?start=2259028a70d6e1a44ad2cfd81955b577a43e8ef6&end=35c5e67c69cbde49b47fe537e296803b6a25b456&stat=instructions:u).

### Overall result: no relevant changes - no action needed


@rustbot label: -perf-regression



### [Instruction count](https://perf.rust-lang.org/compare.html?start=2259028a70d6e1a44ad2cfd81955b577a43e8ef6&end=35c5e67c69cbde49b47fe537e296803b6a25b456&stat=instructions:u)
This benchmark run did not return any relevant results for this metric.

### [Max RSS (memory usage)](https://perf.rust-lang.org/compare.html?start=2259028a70d6e1a44ad2cfd81955b577a43e8ef6&end=35c5e67c69cbde49b47fe537e296803b6a25b456&stat=max-rss)
This benchmark run did not return any relevant results for this metric.

### [Cycles](https://perf.rust-lang.org/compare.html?start=2259028a70d6e1a44ad2cfd81955b577a43e8ef6&end=35c5e67c69cbde49b47fe537e296803b6a25b456&stat=cycles:u)
<details>
<summary>Results</summary>

This is a less reliable metric that may be of interest but was not used to determine the overall result at the top of this comment.

|                                    | mean | range        | count |
|:----------------------------------:|:----:|:------------:|:-----:|
| Regressions ❌ <br /> (primary)    | -    | -            | 0     |
| Regressions ❌ <br /> (secondary)  | 2.2% | [2.2%, 2.2%] | 1     |
| Improvements ✅ <br /> (primary)   | -    | -            | 0     |
| Improvements ✅ <br /> (secondary) | -    | -            | 0     |
| All ❌✅ (primary)                 | -    | -            | 0     |
</details>

### [Binary size](https://perf.rust-lang.org/compare.html?start=2259028a70d6e1a44ad2cfd81955b577a43e8ef6&end=35c5e67c69cbde49b47fe537e296803b6a25b456&stat=size:linked_artifact)
This benchmark run did not return any relevant results for this metric.

**Bootstrap**: 673.288s -> 674.634s (0.20%)
**Artifact size**: 316.02 MiB -> 315.86 MiB (-0.05%)
